Add test to make sure submodule hooks fire

rohan-varma opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #75421

As part of FSDP work, we will be relying on `_register_load_state_dict_pre_hook` to manage some specific logic related to loading state dicts.

This PR adds a test to ensure that _register_load_state_dict_pre_hook can be
used to register hooks on modules that will be used in a nested way, and then
calling load_state_dict on the overall module still calls those hooks
appropriately.

Differential Revision: [D35434726](https://our.internmc.facebook.com/intern/diff/D35434726/)

rohan-varma committed:
Add test to make sure submodule hooks fire

As part of FSDP work, we will be relying on `_register_load_state_dict_pre_hook` to manage some specific logic related to loading state dicts.

This PR adds a test to ensure that _register_load_state_dict_pre_hook can be
used to register hooks on modules that will be used in a nested way, and then
calling load_state_dict on the overall module still calls those hooks
appropriately.

Differential Revision: [D35434726](https://our.internmc.facebook.com/intern/diff/D35434726/)

[ghstack-poisoned]

awgu commented on test/test_nn.py:
```diff
@@ -20602,18 +20602,36 @@ def my_pre_load_hook_with_module(
+                self.mod = mod
+
+        submodule = MyModule()
+        for i, m in enumerate([MyModuleContainer(submodule), submodule]):
```
Curious what the intended purpose of the `i = 1` iteration is. It seems like it is testing that we can register multiple pre-hooks (without modules) and can register multiple pre-hooks with modules. Otherwise, it reduces to not testing a submodule but just a top-level module.

It was not immediately clear to me that `hook_called_offset` is there exactly because we are double-registering hooks on `submodule`. If we want to test multi-registering of hooks, then maybe we should separate it or explain this is what the loop is doing?

rohan-varma committed:
Update on "Add test to make sure submodule hooks fire"


As part of FSDP work, we will be relying on `_register_load_state_dict_pre_hook` to manage some specific logic related to loading state dicts.

This PR adds a test to ensure that _register_load_state_dict_pre_hook can be
used to register hooks on modules that will be used in a nested way, and then
calling load_state_dict on the overall module still calls those hooks
appropriately.

Differential Revision: [D35434726](https://our.internmc.facebook.com/intern/diff/D35434726/)

[ghstack-poisoned]

rohan-varma committed:
Update on "Add test to make sure submodule hooks fire"


As part of FSDP work, we will be relying on `_register_load_state_dict_pre_hook` to manage some specific logic related to loading state dicts.

This PR adds a test to ensure that _register_load_state_dict_pre_hook can be
used to register hooks on modules that will be used in a nested way, and then
calling load_state_dict on the overall module still calls those hooks
appropriately.

Differential Revision: [D35434726](https://our.internmc.facebook.com/intern/diff/D35434726/)

[ghstack-poisoned]

albanD reviewed (approved):
SGTM

rohan-varma committed:
Update on "Add test to make sure submodule hooks fire"


As part of FSDP work, we will be relying on `_register_load_state_dict_pre_hook` to manage some specific logic related to loading state dicts.

This PR adds a test to ensure that _register_load_state_dict_pre_hook can be
used to register hooks on modules that will be used in a nested way, and then
calling load_state_dict on the overall module still calls those hooks
appropriately.

Differential Revision: [D35434726](https://our.internmc.facebook.com/intern/diff/D35434726/)

[ghstack-poisoned]
